abn/watcher/handler.go

```diff
@@ -1,105 +1,306 @@
+	condition *string
+	object    *unstructured.Unstructured
+}
+
+// getApplicationObjects identifies a list of objects related to application based on the name
+func getApplicationObjects(namespace string, application string, appConfig appDetails, informerFactories map[string]dynamicinformer.DynamicSharedInformerFactory) map[string][]trackObject {
+	// initialize
+	var trackToObjectList = map[string][]trackObject{}
+	tracks := trackNames(application, appConfig)
+	for _, track := range tracks {
+		trackToObjectList[track] = []trackObject{}
+	}
+
+	// get objects by resource type
+	for _, r := range appConfig.Resources {
+		lister := informerFactories[namespace].ForResource(r.GroupVersionResource).Lister()
+		objs, err := lister.List(labels.NewSelector())
 		if err != nil {
-			log.Logger.Warn(err)
+			// no such objects; can happen if not deployed
 			continue
 		}
+		// reduce to only those that match expectedObjectNames
+		// all objects are of the same type but for different tracks
 		for _, obj := range objs {
-			wo := watchedObject{Obj: obj.(*unstructured.Unstructured)}
-			if precond(wo) {
-				watchedObjects = append(watchedObjects, wo)
+			ao := obj.(*unstructured.Unstructured)
+			if ao.GetDeletionTimestamp() != nil {
+				// if being deleted, ignore
+				continue
 			}
+			name := ao.GetName()
+			_, ok := trackToObjectList[name]
+			if !ok {
+				// object is not associated with a known track; ignore
+				continue
+			}
+			// create trackObject and add to list of objects for track
+			trackToObjectList[name] = append(trackToObjectList[name], trackObject{gvr: r.GroupVersionResource, condition: r.Condition, object: ao})
 		}
 	}
-	return watchedObjects
+
+	return trackToObjectList
+}
+
+// isTrackReady checks that all expected objects for the track exist, that the version is defined (consistently) and that the objects are ready
+func isTrackReady(track string, trackObjects []trackObject, expectedNumberTrackObjects int) (string, bool) {
+	// all objects exist
+	if len(trackObjects) != expectedNumberTrackObjects {
+		log.Logger.Debugf("expected %d objects; found %d (track: %s)", expectedNumberTrackObjects, len(trackObjects), track)
+		return "", false
+	}
+
+	// single version identified on at least one object
+	var version string
+	for _, to := range trackObjects {
+		v, ok := getVersion(to.object)
+		if ok {
+			if version != "" && version != v {
+				// different versions on resources of the same track
+				log.Logger.Debugf("inconsistent value for label %s (track: %s)", versionLabel, track)
+				return "", false
+			}
+			version = v
+		}
+	}
+	if version == "" {
+		log.Logger.Debugf("no value for label %s found (track: %s)", versionLabel, track)
+		return "", false
+	}
+
+	// all objects are ready
+	for _, to := range trackObjects {
+		if !isObjectReady(to.object, to.gvr, to.condition) {
+			log.Logger.Debugf("no object found of type %v (track: %s)", to.gvr, track)
+			return "", false
+		}
+	}
+
+	return version, true
```

Comment:
i just noticed that the string and the boolean part of this return value correspond to each other. It is always non-empty string/true or empty string/false. I am wondering if it makes sense to just return the string and rename the function.